### PR TITLE
Use grpc client interceptors to properly check for auth requirement

### PR DESCRIPTION
**Pull request by EngHabu**
Signed-off-by: Haytham Abuelfutuh <haytham@afutuh.com>

# TL;DR
Change client auth mechanism to attempt to build a TokenSource whenever any client call received the equivalent of http 401 UNAUTHORIZED response code.

## Type
 - [X] Bug Fix
 - [ ] Feature
 - [ ] Plugin

## Are all requirements met?

 - [X] Code completed
 - [ ] Smoke tested
 - [ ] Unit tests added
 - [ ] Code documentation added
 - [ ] Any pending items have an associated Issue

## Complete description
Existing mechanism to build an authenticated client depends on auth negotiation at the very first time the gRPC client connection is built. If at the time:
1. Auth hasn't been enabled on admin, or
2. It fails to query for Auth endpoints for any transient network reasons.
It'll fallback to assuming Auth isn't enabled on the server and build an unauthenticated gRPC client. Once that's built, it's reused for the entirety of the lifetime of the process.
If auth is enabled (or config changed) at a later point, all existing clients will fail to authenticate and never recover.

With this change, it establishes a better pattern akin to the HTTP-recognized behavior of responding to auth code failures by attempting to authenticate.

After the change, the client will always be built unauthenticated, as soon as the first request is sent and receives an authentication error, it'll attempt to negotiate auth and build a valid token source.

## Tracking Issue

fixes https://github.com/flyteorg/flyte/issues/2841

### CODE REVIEW

**Comment by EngHabu** [comment]
@mayitbeegh @pmahindrakar-oss @katrogan still WIP but please take an initial look at how this is refactored... and provide any early feedback you got. Thank you!

File: clients/go/admin/auth_interceptor.go [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+// It will first invoke the grpc pipeline (to proceed with the request) with no modifications. It's expected for the grpc
+// pipeline to already have a grpc.WithPerRPCCredentials() DialOption. If the tokenSource has already been initialized,
+// it'll take care of refreshing when it expires... etc.
+// If the first invocation fails with an auth error, this interceptor will then attempt to establish a token source once
+// more. It'll fail hard if it couldn't do so (i.e. it will no longer attempt to send an unauthenticated request). Once
+// a token source has been created, it'll invoke the grpc pipeline again, this time the grpc.PerRPCCredentials should
+// be able to find and acquire a valid AccessToken to annotate the request with.
+func newAuthInterceptor(cfg *Config, tokenCache cache.TokenCache, tokenSource *CustomHeaderTokenSource) grpc.UnaryClientInterceptor {
+	return func(ctx context.Context, method string, req, reply interface{}, cc *grpc.ClientConn, invoker grpc.UnaryInvoker, opts ...grpc.CallOption) error {
+		err := invoker(ctx, method, req, reply, cc, opts...)
```
Comment (pmahindrakar-oss) [inline clients/go/admin/auth_interceptor.go:58]: Can we may be log this err too

File: clients/go/admin/auth_interceptor.go [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+	if err != nil {
+		return fmt.Errorf("failed to fetch client metadata. Error: %v", err)
+	}
+
+	tSource, err := tokenSourceProvider.GetTokenSource(ctx)
+	if err != nil {
+		return err
+	}
+
+	tokenSource.Reset(tSource, clientMetadata.AuthorizationMetadataKey, cfg.UseInsecureConnection)
```
Comment (pmahindrakar-oss) [inline clients/go/admin/auth_interceptor.go:42]: we are using tokenSource and tSource which is confusing to read. May be can tokenSource be renamed to singletonTokenSource

**Review by pmahindrakar-oss — COMMENTED** [review]
LGTM

File: clients/go/admin/auth_interceptor.go [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+	return nil
+}
+
+// newAuthInterceptor creates a new grpc.UnaryClientInterceptor that forwards the grpc call and inspects the error.
+// If an auth error is detected, it'll panic the process. This is useful for scenarios when auth is enabled after the client
```
Comment (katrogan) [inline clients/go/admin/auth_interceptor.go:47]: does this actually panic? do we want it to instead of retrying with a rebuilt client?

File: clients/go/admin/auth_interceptor.go (hunk 2 above)
Comment (EngHabu) [reply clients/go/admin/auth_interceptor.go]: plz let me know what do you think now

File: clients/go/admin/auth_interceptor.go (hunk 3 above)
Comment (EngHabu) [reply clients/go/admin/auth_interceptor.go:48]: oh good catch! I guess I was doing some next level DDD (Documentation Driven Development. For a split sec I thought I just coined that but google disappointed me!).

So it does try with the rebuilt client... I initially thought to make it panic to force a restart... but I don't think we need to anymore because it will attempt to rebuild the client on every subsequent call until it manages to get through... so I'll update the docs... plz let me know what you think.

File: clients/go/admin/auth_interceptor_test.go [hunk 4]
```diff
@@ -0,0 +1,205 @@
...
+		f := NewPerRPCCredentialsFuture()
+		interceptor := newAuthInterceptor(&Config{}, &mocks.TokenCache{}, f)
+		otherError := func(ctx context.Context, method string, req, reply interface{}, cc *grpc.ClientConn, opts ...grpc.CallOption) error {
+			return status.New(codes.Canceled, "").Err()
+		}
+
+		assert.Error(t, interceptor(context.Background(), "POST", nil, nil, nil, otherError))
+	})
+
+	t.Run("Unauthenticated first time, succeed", func(t *testing.T) {
```
Comment (katrogan) [inline clients/go/admin/auth_interceptor_test.go:128]: is it possible to add a test for the authenticated code path or is that too hairy for unit tests?

**Review by katrogan — APPROVED** [review]

**Review by katrogan — DISMISSED** [review]

**Review by pmahindrakar-oss — APPROVED** [review]

**Review by katrogan — APPROVED** [review]
